controllers/sandbox_controller.go

```diff
@@ -0,0 +1,155 @@
+type SandboxReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=agents.x-k8s.io,resources=sandboxes,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=agents.x-k8s.io,resources=sandboxes/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=agents.x-k8s.io,resources=sandboxes/finalizers,verbs=update
+//+kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Sandbox object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *SandboxReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	sandbox := &sandboxv1alpha1.Sandbox{}
+	if err := r.Get(ctx, req.NamespacedName, sandbox); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("sandbox resource not found. Ignoring since object must be deleted")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Failed to get sandbox")
+		return ctrl.Result{}, err
+	}
+
+	// Check if the sandbox is being deleted
+	if !sandbox.ObjectMeta.DeletionTimestamp.IsZero() {
+		if controllerutil.ContainsFinalizer(sandbox, sandboxFinalizer) {
+			if err := r.finalizeSandbox(ctx, sandbox); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			controllerutil.RemoveFinalizer(sandbox, sandboxFinalizer)
+			if err := r.Update(ctx, sandbox); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Add finalizer for this CR
+	if !controllerutil.ContainsFinalizer(sandbox, sandboxFinalizer) {
+		controllerutil.AddFinalizer(sandbox, sandboxFinalizer)
+		if err := r.Update(ctx, sandbox); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Check if the pod already exists, if not create a new one
+	found := &corev1.Pod{}
+	err := r.Get(ctx, types.NamespacedName{Name: sandbox.Name, Namespace: sandbox.Namespace}, found)
+	if err != nil && errors.IsNotFound(err) {
+		// Define a new pod
+		pod, err := r.podForSandbox(sandbox)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		log.Info("Creating a new Pod", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
+		if err = r.Create(ctx, pod); err != nil {
+			log.Error(err, "Failed to create new Pod", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
+	} else if err != nil {
+		log.Error(err, "Failed to get Pod")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
```

Comment:
Nit: should we apply if the pod already exists?  e.g. what happens if someone changes the pod image

One for the future I think!

Reply:
SSAing everytime for now. Can add logic for if it exists and is same dont apply. Maybe we need not do it.